substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java

```diff
@@ -571,19 +573,36 @@ public List<Path> getBuilderModulePath() {
         }
 
         private List<Path> createTruffleBuilderModulePath() {
-            List<Path> jars = getJars(rootDir.resolve(Paths.get("lib", "truffle")), "truffle-api", "truffle-runtime", "truffle-enterprise");
+            Path libTruffleDir = rootDir.resolve(Paths.get("lib", "truffle"));
+            List<Path> jars = getJars(libTruffleDir, "truffle-api", "truffle-runtime", "truffle-enterprise");
             if (!jars.isEmpty()) {
                 /*
                  * If Truffle is installed as part of the JDK we always add the builder modules of
                  * Truffle to the builder module path. This is legacy support and should in the
                  * future no longer be needed.
                  */
-                jars.addAll(getJars(rootDir.resolve(Paths.get("lib", "truffle")), "truffle-compiler"));
+                jars.addAll(getJars(libTruffleDir, "truffle-compiler"));
                 Path builderPath = rootDir.resolve(Paths.get("lib", "truffle", "builder"));
                 if (Files.exists(builderPath)) {
                     jars.addAll(getJars(builderPath, "truffle-runtime-svm", "truffle-enterprise-svm"));
+                    if (libJvmciDir != null) {
+                        // truffle-runtime-svm depends on polyglot, which is not part of non-jlinked
+                        // JDKs
+                        jars.addAll(getJars(libJvmciDir, "polyglot"));
```

Comment:
This is most certainly incorrect. Starting with 23.1 polyglot is no longer part of the JDK build and passed by the user on the module-path. We still have some code that picks up jars from the JDK but it will no longer be reachable starting with 23.1.

Reply:
Sorry, but I'm not sure I fully understand what is needed here. I took a guess and now only add `polyglot.jar` if `truffle-runtime-svm` (and the enterprise equivalent) jars are there as well. See https://github.com/oracle/graal/pull/7205/commits/0c518c76252cc597c265bf11916a84f7de415492

@chumer Is that the correct thing to do? If not, what would it be?

> truffle-runtime.svm.jar now gets added using the truffle-svm macro.

Could you clarify what the `truffle-svm` macro is?

Reply:
> Could you clarify what the truffle-svm macro is?

@jerboaa if you build with:
```
mx --primary-suite-path substratevm --components="Native Image,LibGraal,Truffle SVM Macro" --native-images=native-image,lib:jvmcicompiler build
```

notice the `Truffle SVM Macro` part the generated distribution will include the `truffle-svm` macro:

```
❯ cat sdk/latest_graalvm_home/lib/svm/macros/truffle-svm/native-image.properties 
ImageModulePath = ${.}/../../../truffle/builder/truffle-runtime-svm.jar
Args = -H:CLibraryPath=${.}/builder
IgnoreIfBuilderOnClasspath = true
```

You can then use the macro with:
```
native-image --macro:truffle-svm --other-flag my-app
```

Reply:
OK, thanks. It's still not clear to me what **exactly** is being requested by @chumer to get this patch approved :-)

Reply:
I need clarification on why these changes are necessary. You should not need changes in the driver at all. You hardcode many library names here, which should not be done.

The --macro:truffle-svm should *never* be used manually. It's activated implicitly as part of adding truffle-runtime.jar on the module/class-path.